src/test/java/SortOperationTest.java

```diff
@@ -0,0 +1,352 @@
+
+        sortByClauses.add(partSbc);
+        sortByClauses.add(offSbc);
+
+        BatchCollect batchCollect = new BatchCollect("_time", 20, sortByClauses);
+
+        createData(batchCollect, sqlContext);
+
+        List<Row> partition = batchCollect.getCollectedAsDataframe().select("partition").collectAsList();
+        List<Row> offset = batchCollect.getCollectedAsDataframe().select("offset").collectAsList();
+
+        // should be first sorted by partition and then in those partitions sorted by offset
+        Assertions.assertEquals("[3, 3, 3, 3, 3, 2, 2, 2, 2, 2, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0]",
+                Arrays.toString(partition.stream().map(r -> r.getAs(0).toString()).toArray()));
+        Assertions.assertEquals("[4, 3, 2, 1, 0, 4, 3, 2, 1, 0, 4, 3, 2, 1, 0, 4, 3, 2, 1, 0]",
+                Arrays.toString(offset.stream().map(r -> r.getAs(0).toString()).toArray()));
+        Assertions.assertEquals(20, partition.size());
+        Assertions.assertEquals(20, offset.size());
+    }
+
+    @Test
+    public void testTwoSortByClauseDescending_IP() {
+        SparkSession sparkSession = SparkSession.builder().master("local[*]").getOrCreate();
+        SQLContext sqlContext = sparkSession.sqlContext();
+
+        sparkSession.sparkContext().setLogLevel("ERROR");
+
+        ArrayList<SortByClause> sortByClauses = new ArrayList<>();
+        SortByClause partSbc = new SortByClause();
+        partSbc.setFieldName("partition");
+        partSbc.setDescending(true);
+        partSbc.setLimit(20);
+        partSbc.setSortAsType(SortByClause.Type.NUMERIC);
+        SortByClause ipSbc = new SortByClause();
+        ipSbc.setFieldName("IP");
+        ipSbc.setDescending(true);
+        ipSbc.setLimit(20);
+        ipSbc.setSortAsType(SortByClause.Type.IP_ADDRESS);
+
+        sortByClauses.add(partSbc);
+        sortByClauses.add(ipSbc);
+
+        BatchCollect batchCollect = new BatchCollect("_time", 20, sortByClauses);
+
+        createData(batchCollect, sqlContext);
+
+        List<Row> partition = batchCollect.getCollectedAsDataframe().select("partition").collectAsList();
+        List<Row> ip = batchCollect.getCollectedAsDataframe().select("ip").collectAsList();
+
+        // should be first sorted by partition and then in those partitions sorted by IP
+        Assertions.assertEquals("[3, 3, 3, 3, 3, 2, 2, 2, 2, 2, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0]",
+                Arrays.toString(partition.stream().map(r -> r.getAs(0).toString()).toArray()));
+        Assertions.assertEquals("[192.168.1.4, 192.168.1.3, 192.168.1.2, 192.168.1.1, 192.168.1.0, 192.168.1.4, " +
+                        "192.168.1.3, 192.168.1.2, 192.168.1.1, 192.168.1.0, 192.168.1.4, 192.168.1.3, 192.168.1.2, " +
+                        "192.168.1.1, 192.168.1.0, 192.168.1.4, 192.168.1.3, 192.168.1.2, 192.168.1.1, 192.168.1.0]",
+                Arrays.toString(ip.stream().map(r -> r.getAs(0).toString()).toArray()));
```

Comment:
Preferably use 127.x.x.x ip range as 192.168.x.x is routable ip, even if private